### ISSUE TITLE
Type inference fails to exclude null

### ISSUE THREAD

**Issue by pierremarc** [microsoft/TypeScript#42958]
# Bug Report

### 🔎 Search Terms

inference

### 🕗 Version & Regression Information

v4.1.5

### ⏯ Playground Link

[Provided](https://www.typescriptlang.org/play?ts=4.2.0-beta#code/JYOwLgpgTgZghgYwgAgILIN4Chm+XALmRAFcBbAI2mQB9iSAbBgbiwF9WsYSQExgA9iGQBzCGAA8AaWQQAHpBAATAM7IA1hACeAmGgB8ACk1aiUgDTIBFAFZFUASiKlK0bHmQIhKsMgBucAwkKAC8VrYA2iYAujh4wHqGAUEoAIQhYaRMDphxHrhQ4iRQwsnBrB5seciFYMXCAAysVVhAA)

### 💻 Code

```ts
interface A {
    a: number | null;
};

function get<K extends keyof A>(key: K, obj: A): number {
    const value = obj[key];
    if (value !== null) {
      /*
       * Type 'A[K]' is not assignable to type 'number'.
       *   Type 'number | null' is not assignable to type 'number'.
       *     Type 'null' is not assignable to type 'number'.(2322)
       */
        return value;
    }
    return 0;
};
```
      

### 🙁 Actual behavior

`value` is said to be possibly `null`

### 🙂 Expected behavior

`value` should be evaluated as `number`

**Comment by RyanCavanaugh** [comment]
@ahejlsberg if you want a super-concise summary of what you're doing 😉. Otherwise close as dupe

**Comment by ahejlsberg** [comment]
This is effectively a duplicate of #13995 (because `A[K]` is a type variable constrained by a union type `number | null`).

**Comment by ahejlsberg** [comment]
This issue is now fixed in #43183.
